src/optim/FishLeg/fishleg.py

```diff
@@ -278,34 +283,99 @@ def update_aux(self) -> None:
         u_norm = torch.sqrt(u2)
 
-        reg_term = 0.0
-        quad_term = 0.0
-        linear_term = 0.0
+        aux_loss = 0
 
-        p_idx = 0
-        for module in self.model.modules():
-            if not isinstance(module, nn.Sequential) and hasattr(module, "Qv"):
-                layer_grads = []
-                for name, param in module.named_parameters():
-                    if "fishleg_aux" not in name:
-                        layer_grads.append(param.grad.data / u_norm)
+        if self.method == "antithetic":
+            nat_grads = []
+            p_idx = 0
+            for module in self.model.modules():
+                if not isinstance(module, nn.Sequential) and hasattr(module, "Qv"):
+                    layer_grads = []
+                    for name, param in module.named_parameters():
+                        if "fishleg_aux" not in name:
+                            layer_grads.append(param.grad.data)
+
+                    qv, qv_b = module.Qv(layer_grads)
+                    aux_loss -= torch.sum(grads[p_idx] * qv)
+                    aux_loss -= torch.sum(grads[p_idx + 1] * qv_b)
+
+                    aux_loss += damping / 2 * torch.sum(qv**2)
+                    aux_loss += damping / 2 * torch.sum(qv_b**2)
+
+                    nat_grads.append(qv)
+                    nat_grads.append(qv_b)
+                    p_idx += 2
+
+            # Add to gradients
+            p_idx = 0
+            for module in self.model.modules():
+                if not isinstance(module, nn.Sequential) and hasattr(module, "Qv"):
+                    layer_grads = []
+                    for name, param in module.named_parameters():
+                        if "fishleg_aux" not in name:
+                            param.data += self.eps * nat_grads[p_idx]
+                            p_idx += 1
+
+            pred_y = self.model(data_x)
+            plus_loss = self.likelihood.nll(pred_y, samples_y)
+
+            # Minus gradients
+            p_idx = 0
+            for module in self.model.modules():
+                if not isinstance(module, nn.Sequential) and hasattr(module, "Qv"):
+                    layer_grads = []
+                    for name, param in module.named_parameters():
+                        if "fishleg_aux" not in name:
+                            param.data -= 2 * self.eps * nat_grads[p_idx]
+                            p_idx += 1
+
+            pred_y = self.model(data_x)
+
+            minus_loss = self.likelihood.nll(pred_y, samples_y)
+
+            # Restore parameters
+            p_idx = 0
+            for module in self.model.modules():
+                if not isinstance(module, nn.Sequential) and hasattr(module, "Qv"):
+                    layer_grads = []
+                    for name, param in module.named_parameters():
+                        if "fishleg_aux" not in name:
+                            param.data += self.eps * nat_grads[p_idx]
```

Comment:
We should also have a chat about this. We need to choose between memory and computational speed. I feel a bit torn.

Reply:
Sure, yes I was also thinking this. I went this way first since it is simpler and also the memory capacity is a hard limit, whereas time is less restrictive. I think having an option for parallelising this could be good though. There may also be other ways to structure the computation which may help.